### ISSUE TITLE
updating chans in handlers does not work

### ISSUE THREAD

**Issue by Ptival** [UCSD-PL/kraken#63]
The problem is that this line of code:

```
curtab := spawn
```

compiles down to:

```
curtab <- exec Tab (str_of_string "/home/ucsd/vcr/python-browser-8/tab.py")
(tr ~~~ expand_ktrace (KRecv c (Navigate url) :: tr))
<@> [In curtab comps] * [In output comps] * all_bound comps * [In c comps] * (tr ~~ [KInvariant kst]) * emp;
```

which capturs curtab in such a way that the goal [In curtab comps] now refers to the new curtab which is actually _not_ in comps.

---

@ztatlock suggested a fix that I am going to try now: we need to update comps to (curtab :: comps) from that point on.

**Comment by Ptival** [comment]
The fix for that part of the code actually works quite well. However, it unveils a more complicated problem in ValidExchange. Consider the following code:

```
curtab = spawn(Tab)
curtab = spawn(Tab)
send(curtab, Go(url))
```

and the following generated code:

```
Let curtab := curtab s.

[...]

| VE_Tab_Navigate_0 :
    forall c url,
      True ->
      ValidExchange (mkst
        (curtab :: comps)
        (tr ~~~ KSend curtab (Go (url))
          :: KExec (str_of_string "/home/ucsd/vcr/python-browser-8/tab.py") curtab
          :: KExec (str_of_string "/home/ucsd/vcr/python-browser-8/tab.py") curtab
          :: KRecv c (Navigate url) :: tr)
        (curtab)
        (output)
      )
```

One can see two problems:
1. curtab should be quantified in that constructor, as it is not talking about (curtab s), but rather a new variable named curtab.
2. the two occurences of curtab in each KExec should actually be distinguished.

---

This makes the problem of determining the variables to quantify in the constructor quite complicated, as we might need more variable names than there are in the program...

Basically, we would want to go from:

```
curtab = spawn(Tab)
curtab = spawn(Tab)
send(curtab, Go(url))
```

to:

```
| VE_Tab_Navigate_0 :
    forall c url curtab0 curtab1,
      True ->
      ValidExchange (mkst
        (curtab1 :: comps)
        (tr ~~~ KSend curtab1 (Go (url))
          :: KExec (str_of_string "/home/ucsd/vcr/python-browser-8/tab.py") curtab1
          :: KExec (str_of_string "/home/ucsd/vcr/python-browser-8/tab.py") curtab0
          :: KRecv c (Navigate url) :: tr)
        (curtab1)
        (output)
      )
```

You might notice this is crazy...

**Comment by Ptival** [comment]
Trying to push this further, I devised:

```
tmp = c1;
c1 = c2;
c2 = tmp;
send(c2, M("Secret"));
```

which generates:

```
Let c1 := c1 s.
Let c2 := c2 s.

Inductive ValidExchange : kstate -> Prop :=
  | VE_Comp_M_0 :
    forall c s,
      True ->
      ValidExchange (mkst
        (comps)
        (tr ~~~ KSend c2 (M ((str_of_string "Secret"))) :: KRecv c (M s) :: tr)
        (c2)
        (c1)
      )
```

As you can see, the ValidExchange is bogus: it allows to send the message to the second component of the initial state, while the code effectively sends the message to the first component of the initial state.

**Comment by Ptival** [comment]
What we could do is the following:

```
Inductive ValidExchange : kstate -> Prop :=
| VE_Comp_M_0 :
forall c s,
  True ->
  ValidExchange (mkst
                   (comps)
                   (tr ~~~
                       (
                         let tmp := c1 in
                         let c1 := c2 in
                         let c2 := tmp in
                         KSend c2 (M ((str_of_string "Secret")))
                       ) :: KRecv c (M s) :: tr)
                   (c2)
                   (c1)
                )
```

That is, we cripple the trace with all the lets, so that things are well-defined where they need to.
The proofs for this example all go through with that change.

What think?

**Comment by Ptival** [comment]
Actually, that will be annoying since (as we figured out last time) the lets scopes are inverted with regards to the trace.

**Comment by ztatlock** [comment]
Don't we just need to update ktrace_spec as well as ktrace_impl?

**Comment by Ptival** [comment]
that doesn't really work because of the let bindings.

**Comment by ztatlock** [comment]
Ah, I see.

So are we gonna use your let idea then?

**Comment by Ptival** [comment]
Dunno yet. The problem is that we want this:

```
     a     ::     b     ::     c     ::     d     ::     nil
  <---------- one binding ---------->
  <------ another ------->
  <- etc. ->
```

It is not quite possible, even syntactically...

I might have an idea, will try to see how it goes once at school.
